fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistryEvents.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.event.registry;
+
+import java.util.Collection;
+import java.util.List;
+
+import com.mojang.serialization.Codec;
+import org.jetbrains.annotations.ApiStatus;
+
+import net.minecraft.registry.Registry;
+import net.minecraft.registry.RegistryKey;
+import net.minecraft.registry.RegistryLoader;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.fabricmc.fabric.impl.registry.sync.DynamicRegistryRegistrationContextImpl;
+
+/**
+ * Events relating to dynamic registries.
+ */
+public final class DynamicRegistryEvents {
+	// Events
+
+	/**
+	 * An event for registering new dynamic registries.
+	 */
+	public static final Event<RegisterRegistries> REGISTER_REGISTRIES = EventFactory.createArrayBacked(RegisterRegistries.class, callbacks -> context -> {
+		for (RegisterRegistries callback : callbacks) {
+			callback.registerDynamicRegistries(context);
+		}
+	});
+
+	// TODO: is this good to have?
+	/**
+	 * An event that is invoked before dynamic registries are loaded.
+	 * An alias for {@link DynamicRegistrySetupCallback#EVENT}.
+	 */
+	public static final Event<DynamicRegistrySetupCallback> BEFORE_LOAD = DynamicRegistrySetupCallback.EVENT;
+
+	// Phases
+
+	/**
+	 * A phase of {@link #REGISTER_REGISTRIES} when vanilla dynamic registries are registered.
+	 * This phase executes before {@linkplain Event#DEFAULT_PHASE the default phase}.
+	 */
+	public static final Identifier REGISTER_VANILLA_REGISTRIES_PHASE = new Identifier("fabric-api", "register_vanilla_registries");
+
+	static {
+		REGISTER_REGISTRIES.addPhaseOrdering(REGISTER_VANILLA_REGISTRIES_PHASE, Event.DEFAULT_PHASE);
+		REGISTER_REGISTRIES.register(REGISTER_VANILLA_REGISTRIES_PHASE, context -> {
+			// Add all vanilla dynamic registries
+			context.add(RegistryLoader.DYNAMIC_REGISTRIES);
+		});
+	}
+
+	private DynamicRegistryEvents() {
+	}
+
+	/**
+	 * Collects all entries for dynamic registries.
+	 *
+	 * @return the entries
+	 */
+	public static List<RegistryLoader.Entry<?>> collectDynamicRegistries() {
```

Comment:
Does this need to be integrated with the datagen API?

Reply:
Wdym integrated? This PR makes the data generation API use this method instead of the vanilla field.

Reply:
Ah right, missed that.